### PR TITLE
Allow creating private pools

### PR DESCRIPTION
**Pull request by MiksuJak**

### CODE REVIEW

File: contracts/truefi2/PoolFactory.sol
Comment (vanruch) [inline contracts/truefi2/PoolFactory.sol:70]: I think we should never change event names. We have them to have a clear history of what was happening with the contract and changing it could be confusing. On the other hand it sucks to have AllowedStatusChanged and BorrowerAllowedStatus changed. Maybe we could use the whitelist term for borrowers instead of Allowance?

File: contracts/truefi2/TrueFiPool2.sol
Comment (vanruch) [inline contracts/truefi2/TrueFiPool2.sol:107]: I think argument name "name" is misleading. I understand that the pool will be called something like "TrueFi Alameda USDC". I have no idea what it should be called. Maybe we need to have separate initializers for public and private pools? This one will stay same and private initializer will take full name and symbol?

**Review by yuchenlintt — CHANGES_REQUESTED** [review]
Looks like this is the right way to implement pools for single borrowers.

I'm not sure I like the idea of calling it `private` though. `private` is an overloaded term that can refer to code visibility (in CS), protection from surveillance (in human rights), or business vs state sectors (in government).

Any suggestions for a better name? I'm considering: `singleBorrowerPool`, `personalPool`, `monopsonyPool`.

File: contracts/truefi2/PoolFactory.sol
Comment (barrutko) [inline contracts/truefi2/PoolFactory.sol:64]: Not sure if we want to rename the event.

File: contracts/truefi2/PoolFactory.sol
Comment (barrutko) [inline contracts/truefi2/PoolFactory.sol:158]: Why empty string?

**Review by barrutko — COMMENTED** [review]
Let's keep old naming unchanged, only add new stuff.

**Review by barrutko — COMMENTED** [review]
Token allowance, borrowers whitelist

**Comment by yuchenlintt** [comment]
Thanks, I think this looks almost ready.

Two remaining questions:
1. Can we come up with a better name than private? I've polled the team in #truefi to see if they have better ideas.
2. Should we restrict single borrower pool creation? Currently we let the public create a pool for any {whitelisted borrower} x {allowed token} combination. Is there a situation where we want to let Alice have a USDC pool but don't want anyone to give her a USDT pool? I'm not sure there is... but if there is such a situation then here are some other options to consider:
  - drop the borrower whitelist and make single borrower pool creation onlyOwner
  - keep the whitelist, but only allow borrowers to create their own single borrower pools

**Comment by MiksuJak** [comment]
> Thanks, I think this looks almost ready.
> 
> Two remaining questions:
> 
> 1. Can we come up with a better name than private? I've polled the team in #truefi to see if they have better ideas.
> 2. Should we restrict single borrower pool creation? Currently we let the public create a pool for any {whitelisted borrower} x {allowed token} combination. Is there a situation where we want to let Alice have a USDC pool but don't want anyone to give her a USDT pool? I'm not sure there is... but if there is such a situation then here are some other options to consider:
> 
> * drop the borrower whitelist and make single borrower pool creation onlyOwner
> * keep the whitelist, but only allow borrowers to create their own single borrower pools

I personally belive, that letting the protocol go free is generally a good practice. As long as creation of a new pool does not hurt us in any way, we're fine I guess. That might create new opportunities for us to pursue. If someone creates a new pool and users willingly join it, maybe it is the right move to start supporting such pool that has a hight interest. We still can decide not to support the pool, do not create a farm and do not whitelist it in liquidator and it would kind of just stay there on it's own. 

That's just my opinion. Waiting for some feedback from the rest of truefi team :)

**Comment by MiksuJak** [comment]
By the way, we probably need to change liquidator, so it whitelists pools instead of tokens. I think this PR is big enough. If it goes through we should do it in a new one.

**Comment by yuchenlintt** [comment]
Agreed, let's keep this PR focused on pool creation.

1. My poll came back with 11 unanimous votes for `singleBorrowerPool`, so let's go with that.
2. I completely agree with you. Decentralization is better, creating pools is harmless, and it would be a strong sign of lending interest if lenders are depositing into a pool with no farm.

File: contracts/truefi2/PoolFactory.sol
Comment (hal909) [inline contracts/truefi2/PoolFactory.sol:172]: I'm not sure if we want whitelisted borrowers to be able to make a pool for every token. I will get clarification on this from the team. @yuchenlintt

**Review by hal909 — CHANGES_REQUESTED** [review]
Just need clarification on whether whitelisted borrowers should be able to create multiple pools. Once I get this, I can approve

File: contracts/truefi2/PoolFactory.sol
Comment (yuchenlintt) [inline contracts/truefi2/PoolFactory.sol:44]:
```suggestion
    // @dev Mapping of borrowers to mapping of ERC20 token's addresses to its single borrower pool
```

File: contracts/truefi2/PoolFactory.sol
Comment (yuchenlintt) [inline contracts/truefi2/PoolFactory.sol:58]:
```suggestion
     * @dev Event to show creation of the new single borrower pool
```

File: contracts/truefi2/PoolFactory.sol
Comment (yuchenlintt) [inline contracts/truefi2/PoolFactory.sol:73]:
```suggestion
     * @dev Event to show that borrower is now allowed/disallowed to have a single borrower pool
```

File: contracts/truefi2/PoolFactory.sol
Comment (yuchenlintt) [inline contracts/truefi2/PoolFactory.sol:164]:
```suggestion
     * @dev Create a new single borrower pool behind proxy. Update new pool's implementation.
```

File: contracts/truefi2/TrueFiPool2.sol
Comment (yuchenlintt) [inline contracts/truefi2/TrueFiPool2.sol:121]:
```suggestion
    function initializeSingleBorrower(
```

File: contracts/truefi2/interface/ITrueFiPool2.sol
Comment (yuchenlintt) [inline contracts/truefi2/interface/ITrueFiPool2.sol:18]:
```suggestion
    function initializeSingleBorrower(
```

**Review by yuchenlintt — CHANGES_REQUESTED** [review]

**Review by yuchenlintt — APPROVED** [review]
I'm changing my review to an Approve since the PR itself looks good to me now. Still waiting on a higher-up decision on whether we want to launch this, but there's no need to block on me if Hal gets that thumbs up.

**Review by hal909 — APPROVED** [review]

**Pull request merged**
